### PR TITLE
Added unit and Integration Tests for windows azure storage plugin

### PR DESCRIPTION
**Pull request by arroyc**

### CODE REVIEW

File: src/test/java/IntegrationTests/IntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        public final String storageCredId;
+
+        TestEnvironment() {
+            
+            azureStorageAccountName = TestEnvironment.loadFromEnv("AZURE_STORAGE_TEST_STORAGE_ACCOUNT_NAME");
+            azureStorageAccountKey1 = TestEnvironment.loadFromEnv("AZURE_STORAGE_TEST_STORAGE_ACCOUNT_KEY1");
+            azureStorageAccountKey2 = TestEnvironment.loadFromEnv("AZURE_STORAGE_TEST_STORAGE_ACCOUNT_KEY2");
+            /*
+            azureStorageAccountName = "testazstorageplugin";
+            azureStorageAccountKey1 = "itSz2vgC33RGMPYYwucTpxgpfXXTsBqUCst/Y6jGOPTzHCUhAGqdfCz83xnr7tKuYMqN3eN40ev9q40HXA/Tjg==";
```
Comment (clguimanMSFT) [inline src/test/java/IntegrationTests/IntegrationTest.java]: please reset the keys :)

File: src/test/java/IntegrationTests/WAStorageClientIT.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        }
+    }
+    
+    @After
+    public void tearDown() {
+        for (int k = 0; k < uploadFile.size(); k++) {
+            uploadFile.get(k).delete();
+        }
+        uploadFile.clear();
+        System.gc();
```
Comment (clguimanMSFT) [inline src/test/java/IntegrationTests/WAStorageClientIT.java]: why are you running the garbage collection?

File: src/test/java/IntegrationTests/IntegrationTest.java (hunk 1 above)
Comment (arroyc) [inline src/test/java/IntegrationTests/IntegrationTest.java]: done

**Comment by arroyc** [comment]
@clguimanMSFT  .. it's working and PR is updated  👍

File: src/test/java/IntegrationTests/IntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
```
Comment (clguimanMSFT) [inline src/test/java/IntegrationTests/IntegrationTest.java]: either add a license header or remove this comment

File: src/test/java/IntegrationTests/IntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+        TestEnvironment() {
+            azureStorageAccountName = TestEnvironment.loadFromEnv("AZURE_STORAGE_TEST_STORAGE_ACCOUNT_NAME");
+            azureStorageAccountKey1 = TestEnvironment.loadFromEnv("AZURE_STORAGE_TEST_STORAGE_ACCOUNT_KEY1");
+            azureStorageAccountKey2 = TestEnvironment.loadFromEnv("AZURE_STORAGE_TEST_STORAGE_ACCOUNT_KEY2");
+            
+            blobURL = Utils.DEF_BLOB_URL;
+            AzureCredentials.StorageAccountCredential u = new AzureCredentials.StorageAccountCredential(azureStorageAccountName, azureStorageAccountKey1, blobURL);
+            sampleStorageAccount = new StorageAccountInfo(azureStorageAccountName,azureStorageAccountKey1, blobURL);
+            storageCredId = u.getId();
```
Comment (clguimanMSFT) [inline src/test/java/IntegrationTests/IntegrationTest.java]: seems that you're not using the storageCredId

File: src/test/java/IntegrationTests/WAStorageClientDownloadIT.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+public class WAStorageClientDownloadIT extends IntegrationTest{
+    
+    private static final Logger LOGGER = Logger.getLogger(WAStorageClientDownloadIT.class.getName());
+    
+    /**
+     * Test of validateStorageAccount method, of class WAStorageClient.
+     */
+    @Test
+    public void testValidateStorageAccount() throws Exception {
+        System.out.println("validateStorageAccount");
```
Comment (clguimanMSFT) [inline src/test/java/IntegrationTests/WAStorageClientDownloadIT.java]: you should also have a negative test

File: src/test/java/IntegrationTests/WAStorageClientDownloadIT.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        StorageAccountInfo storageAccount = testEnv.sampleStorageAccount;
+        boolean expResult = true;
+        boolean result = WAStorageClient.validateStorageAccount(storageAccount);
+        assertEquals(expResult, result);
+    }
+
+    @Test(timeout=1200000)
+    public void testDownloadfromContainer() {
```
Comment (clguimanMSFT) [inline src/test/java/IntegrationTests/WAStorageClientDownloadIT.java]: no need for timeout, the DevOpsCI handles that now

File: src/test/java/IntegrationTests/WAStorageClientUploadIT.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    private static final Logger LOGGER = Logger.getLogger(WAStorageClientUploadIT.class.getName());
+    
+    /**
+     * Test of validateStorageAccount method, of class WAStorageClient.
+     */
+    @Test
+    public void testValidateStorageAccount() throws Exception {
+        System.out.println("validateStorageAccount");
+        StorageAccountInfo storageAccount = testEnv.sampleStorageAccount;
+        boolean expResult = true;
```
Comment (clguimanMSFT) [inline src/test/java/IntegrationTests/WAStorageClientUploadIT.java]: why do you have the same test case in two different classes?

File: src/test/java/IntegrationTests/WAStorageClientDownloadIT.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+            CloudStorageAccount account = new CloudStorageAccount(new StorageCredentialsAccountAndKey(testEnv.azureStorageAccountName, testEnv.azureStorageAccountKey1));
+            CloudBlobClient blobClient = account.createCloudBlobClient();
+            String containerName = "testdownload"+TestEnvironment.GenerateRandomString(15);
+            File directory = new File(containerName);
+            directory.mkdir();
+            container = blobClient.getContainerReference(containerName);
+            container.deleteIfExists();
+            container.createIfNotExists();
+            for(int i=0; i<testEnv.TOTAL_FILES;i++){
+                testEnv.downloadFileList.add(new File(directory, "download"+UUID.randomUUID().toString()+".txt"));
```
Comment (clguimanMSFT) [inline src/test/java/IntegrationTests/WAStorageClientDownloadIT.java]: this part seems very similar with the upload. I would move it in a method inside IntegrationTest class

File: src/test/java/IntegrationTests/WAStorageClientDownloadIT.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            }                         
+            //try to download the same file     
+            Run mockRun = mock(Run.class);
+            Launcher mockLauncher = mock(Launcher.class);
+            WAStorageClient mockStorageClient = spy(WAStorageClient.class);
+            
+            FilePath workspace= new FilePath(mockLauncher.getChannel(), FilenameUtils.getFullPathNoEndSeparator(testEnv.downloadFileList.get(0).getAbsolutePath()));
+            int totalFiles = mockStorageClient.download(mockRun, mockLauncher, TaskListener.NULL, testEnv.sampleStorageAccount, "*.txt", ",archive.zip", "", false, workspace, containerName);
+            
+            assertEquals(testEnv.downloadFileList.size(),totalFiles);
```
Comment (clguimanMSFT) [inline src/test/java/IntegrationTests/WAStorageClientDownloadIT.java]: you should also check the integrity of the files

File: src/test/java/IntegrationTests/WAStorageClientDownloadIT.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+                CloudBlockBlob blob = container.getBlockBlobReference(testEnv.downloadFileList.get(i).getName());
+                blob.uploadText(TestEnvironment.GenerateRandomString(20), "UTF-8", AccessCondition.generateEmptyCondition(), null, null);
+            }                         
+            //try to download the same file     
+            Run mockRun = mock(Run.class);
+            Launcher mockLauncher = mock(Launcher.class);
+            WAStorageClient mockStorageClient = spy(WAStorageClient.class);
+            
+            FilePath workspace= new FilePath(mockLauncher.getChannel(), FilenameUtils.getFullPathNoEndSeparator(testEnv.downloadFileList.get(0).getAbsolutePath()));
+            int totalFiles = mockStorageClient.download(mockRun, mockLauncher, TaskListener.NULL, testEnv.sampleStorageAccount, "*.txt", ",archive.zip", "", false, workspace, containerName);
```
Comment (clguimanMSFT) [inline src/test/java/IntegrationTests/WAStorageClientDownloadIT.java]: why are you excluding archive.zip? I guess you should also have a test to make sure the exclude works

File: src/test/java/IntegrationTests/WAStorageClientDownloadIT.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        }catch(Exception e){
+            LOGGER.log(Level.SEVERE, null, e);
+            Assert.assertTrue(e.getMessage(), false);
+        }
+    }
+    
+    @After
+    public void tearDown() {
+        System.gc();
+        for (int k = 0; k < testEnv.downloadFileList.size(); k++) {
```
Comment (clguimanMSFT) [inline src/test/java/IntegrationTests/WAStorageClientDownloadIT.java]: if you have a tearDown method I would expect a setUp method as well (the temporary file creation should be there

File: src/test/java/IntegrationTests/WAStorageClientDownloadIT.java (hunk 5 above)
Comment (arroyc) [inline src/test/java/IntegrationTests/WAStorageClientDownloadIT.java]: negative tests are already there in unit testing

**Pull request merged**